executor/executor_join.go

```diff
@@ -521,32 +666,109 @@ func (e *HashSemiJoinExec) Next() (*Row, error) {
 				return nil, errors.Trace(err)
 			}
 		}
-		isNull := false
 		if matched {
-			matched, isNull, err = e.rowIsMatched(bigRow)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
+			return bigRow, err
 		}
-		if !matched && e.smallTableHasNull {
-			isNull = true
+	}
+}
+
+func (e *HashSemiJoinExec) doJoin(bigRow *Row) ([]*Row, error) {
+	matched, isNull, err := e.rowIsMatched(bigRow)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !matched && e.smallTableHasNull {
+		isNull = true
+	}
+	if e.anti && !isNull {
+		matched = !matched
+	}
+	// For the auxMode subquery, we return the row with a Datum indicating if it's a match,
+	// For the non-auxMode subquery, we return the matching row only.
+	if e.auxMode {
+		if isNull {
+			bigRow.Data = append(bigRow.Data, types.NewDatum(nil))
+		} else {
+			bigRow.Data = append(bigRow.Data, types.NewDatum(matched))
 		}
-		if e.anti && !isNull {
-			matched = !matched
+		return []*Row{bigRow}, nil
```

Comment:
Can we add an interface named doSemiJoin in joinExec? So we do not have to create a slice here.